rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,777 @@
+        if let Some(state) = self.de_state {
+            core::sc_detect_engine_state_free(state);
+        }
+    }
+}
+
+impl Drop for MQTTTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+pub struct MQTTState {
+    tx_id: u64,
+    request_buffer: Vec<u8>,
+    response_buffer: Vec<u8>,
+    pub protocol_version: u8,
+    transactions: Vec<MQTTTransaction>,
+    connected: bool,
+}
+
+impl MQTTState {
+    pub fn new() -> Self {
+        Self {
+            tx_id: 0,
+            request_buffer: Vec::new(),
+            response_buffer: Vec::new(),
+            protocol_version: 0,
+            transactions: Vec::new(),
+            connected: false,
+        }
+    }
+
+    fn free_tx(&mut self, tx_id: u64) {
+        let len = self.transactions.len();
+        let mut found = false;
+        let mut index = 0;
+        for i in 0..len {
+            let tx = &self.transactions[i];
+            if tx.tx_id == tx_id + 1 {
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&MQTTTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == tx_id + 1 {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    pub fn get_tx_by_pkt_id(&mut self, pkt_id: u32) -> Option<&mut MQTTTransaction> {
+        for tx in &mut self.transactions {
+            if let Some(mpktid) = tx.pkt_id {
+                if mpktid == pkt_id {
+                    return Some(tx);
+                }
+            }
+        }
+        return None;
+    }
+
+    fn new_tx(&mut self, msg: MQTTMessage, toclient: bool) -> MQTTTransaction {
+        let mut tx = MQTTTransaction::new(msg);
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        if toclient {
+            tx.toclient = true;
+        } else {
+            tx.toserver = true;
+        }
+        return tx;
+    }
+
+    fn handle_msg(&mut self, msg: MQTTMessage, toclient: bool) {
+        match msg.op {
+            MQTTOperation::CONNECT(ref conn) => {
+                self.protocol_version = conn.protocol_version;
+                if let Some(tx) = self.get_tx_by_pkt_id(MQTT_CONNECT_PKT_ID) {
+                    &mut MQTTState::set_event(tx, MQTTEvent::DoubleConnect);
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    tx.pkt_id = Some(MQTT_CONNECT_PKT_ID);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::PUBLISH(ref publish) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
```

Comment:
Why return ?

Reply:
Well,  if no `CONNECT` was seen before any other operation, we can simply set the corresponding event and skip this message. However, now that I think about it, you are right -- if we arrived here, then something must have been parsed which is syntactically correct, so might as well store it... _and_ set the event as well. It might still be interesting to see.

Reply:
Removing the `return`.